Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm

```diff
@@ -65,6 +69,122 @@
     return nullString();
 }
 
+void WebExtensionController::getDataRecords(OptionSet<WebExtensionDataType> types, CompletionHandler<void(Vector<Ref<WebExtensionDataRecord>>)>&& completionHandler)
+{
+    if (!m_configuration->storageIsPersistent() || types.isEmpty()) {
+        completionHandler({ });
+        return;
+    }
+
+    auto recordHolder = RecordHolder::create();
+    auto aggregator = MainRunLoopCallbackAggregator::create([recordHolder, completionHandler = WTFMove(completionHandler)]() mutable {
+        Vector<Ref<WebExtensionDataRecord>> records;
+        for (auto entry : recordHolder->recordsMap)
+            records.append(entry.value);
+
+        completionHandler(records);
+    });
+
+    auto uniqueIdentifiers = FileSystem::listDirectory(m_configuration->storageDirectory());
+    for (auto& uniqueIdentifier : uniqueIdentifiers) {
+        auto storageDirectory = FileSystem::pathByAppendingComponent(m_configuration->storageDirectory(), uniqueIdentifier);
+        auto stateFilePath = FileSystem::pathByAppendingComponent(storageDirectory, "State.plist"_s);
```

Comment:
```c++
"State.plist"_s
```
This should likely be in a static getter method on `WebExtensionContext` so it does not expose the file name here (and below) and we can change it in `WebExtensionContext` freely if needed.

You can make a private helper method on `WebExtensionController` to get the state path in one shot for a unique identifier, so you don't have to do this in a couple places in this patch.